### PR TITLE
Implement HNS + tests.

### PR DESCRIPTION
**Pull request by m-kostrzewa**
HNS wrapper + tests. Some values hardcoded for now. Works on virtual test env.

### CODE REVIEW

File: driver/driver.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	d := &ContrailDriver{
+		HnsID: hnsID,
+	}
+	return d, nil
+}
+
+func (d *ContrailDriver) Teardown() error {
+	err := DeleteHNSNetwork(d.HnsID)
```
Comment (Attil) [inline driver/driver.go]: I think it should be logged like other methods.

File: driver/driver.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	// HNSNetworkName is a constant name for HNS network
+	NetworkHNSname = "ContrailHNSNet"
+)
+
+type ContrailDriver struct {
+	HnsID string
+}
+
+func NewDriver() (*ContrailDriver, error) {
+
```
Comment (Attil) [inline driver/driver.go]: Logging data used to create the driver is a good idea.

File: driver/driver.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+func (d *ContrailDriver) Join(req *network.JoinRequest) (*network.JoinResponse, error) {
+	logrus.Debugln("=== Join")
+	logrus.Debugln(req)
+	logrus.Debugln("options:")
+	for k, v := range req.Options {
+		fmt.Printf("%v: %v\n", k, v)
+	}
+	r := &network.JoinResponse{}
+	r.DisableGatewayService = true
```
Comment (Attil) [inline driver/driver.go]: Why?

File: main.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+func main() {
+	d, err := driver.NewDriver()
+	if err != nil {
+		logrus.Error(err)
+	}
+	h := network.NewHandler(d)
+
+	config := sdk.WindowsPipeConfig{
+		// This will set permissions for Everyone user allowing him to open, write, read the pipe
+		SecurityDescriptor: "S:(ML;;NW;;;LW)D:(A;;0x12019f;;;WD)",
```
Comment (Attil) [inline main.go]: Please incorporate SD changes from #7 .

File: tests/driver_test.go [hunk 5]
```diff
@@ -78,36 +149,65 @@ var _ = Describe("Network Driver", func() {
...
 				It("responds with err", func() {})
 			})
-			// Dunno. in standard windows drivers, it's an empty func otherwise
 		})
 
 		Context("on Leave", func() {
+
 			Context("queried endpoint exists", func() {
-				It("responds with proper JoinResponse", func() {}) // nil maybe?
+				It("responds with proper JoinResponse, nil", func() {}) // nil maybe?
```
Comment (Attil) [inline tests/driver_test.go]: That comment is no longer needed

File: tests/hns_test.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+			nets, err := driver.ListHNSNetworks()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(nets)).To(Equal(originalNumNetworks))
+		})
+
+		AfterEach(func() {
+			nets, err := driver.ListHNSNetworks()
+			Expect(err).ToNot(HaveOccurred())
+			for _, n := range nets {
+				if strings.Contains(n.Name, "nat") {
```
Comment (Attil) [inline tests/hns_test.go:119]: Why is that?

**Review by Attil — CHANGES_REQUESTED** [review]

File: driver/driver.go (hunk 1 above)
Comment (m-kostrzewa) [reply driver/driver.go]: HNS log requests as using debugln in hns.go, errors are caught and logged in func main. You mean something different?

File: tests/hns_test.go (hunk 6 above)
Comment (m-kostrzewa) [reply tests/hns_test.go:119]: docker creates its own nat network on startup, we don't want to delete it in HNS as it breaks docker.

File: driver/driver.go (hunk 2 above)
Comment (m-kostrzewa) [reply driver/driver.go]: see below

**Comment by m-kostrzewa** [comment]
applied changes

File: driver/driver.go (hunk 1 above)
Comment (Attil) [reply driver/driver.go]: I meant add a simple information about call to the function like

`logrus.Debugln("=== DeleteEndpoint")`

in `DeleteEndpoint`

**Comment by Attil** [comment]
About `DisableGatewayService` - have you tested whether it still works without it?

**Comment by m-kostrzewa** [comment]
seems to work, dunno why I set it

**Review by Attil — APPROVED** [review]

**Pull request merged**